File: bonnie/registry.go
package bonnie

import (
	"context"
	"log/slog"
	"sync"
	"time"
)

// Agent is the minimal agent record held in the registry. Callers
// persist richer rows in their own DB; this is the shared shape that
// flows through RegistryStore.
type Agent struct {
	ID         string
	Name       string
	URL        string
	Token      string
	Status     string
	LastSeenAt time.Time
}

// RegistryStore abstracts persistent storage of agents. KARR, KITT, and
// DEVON each implement this against their own sqlc-generated queries —
// the package does not know or care about their schemas.
type RegistryStore interface {
	// List returns every agent known to the store.
	List(ctx context.Context) ([]Agent, error)
	// UpdateStatus marks an agent online/offline with the timestamp of
	// the most recent health check.
	UpdateStatus(ctx context.Context, id, status string, lastSeenAt time.Time) error
}

// Status tags recorded by the registry health loop.
const (
	StatusOnline  = "online"
	StatusOffline = "offline"
)

// entry pairs an Agent with its live Client.
type entry struct {
	agent  Agent
	client Client
}

// Registry manages a set of BONNIE clients, one per registered agent,
// and polls them for health in a background goroutine.
type Registry struct {
	store        RegistryStore
	pollInterval time.Duration
	logger       *slog.Logger
	clientOpts   []Option

	mu      sync.RWMutex
	entries map[string]*entry
}

// NewRegistry constructs a Registry that reads agents from store and
// polls them every pollInterval. clientOpts are forwarded to every
// constructed Client so callers can share HTTP clients, loggers, retry
// counts, etc.
//
// pollInterval <= 0 disables automatic polling; callers can still invoke
// Reload and Poll explicitly.
func NewRegistry(store RegistryStore, pollInterval time.Duration, logger *slog.Logger, clientOpts ...Option) *Registry {
	if logger == nil {
		logger = slog.Default()
	}
	return &Registry{
		store:        store,
		pollInterval: pollInterval,
		logger:       logger,
		clientOpts:   clientOpts,
		entries:      map[string]*entry{},
	}
}

// Start kicks off the background health loop. It first reloads from the
// store, then polls on the configured interval until ctx is cancelled.
// It is safe to call Start more than once — subsequent calls each spawn
// a goroutine, so callers typically invoke it exactly once during
// service bootstrap.
func (r *Registry) Start(ctx context.Context) {
	if err := r.Reload(ctx); err != nil {
		r.logger.Warn("bonnie: registry initial reload failed", "error", err)
	}
	if r.pollInterval <= 0 {
		return
	}
	go r.loop(ctx)
}

func (r *Registry) loop(ctx context.Context) {
	t := time.NewTicker(r.pollInterval)
	defer t.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-t.C:
			r.Poll(ctx)
		}
	}
}

// Reload replaces the in-memory set of agents with the rows currently in
// the store. Existing Client instances are reused when URL and Token
// haven't changed.
func (r *Registry) Reload(ctx context.Context) error {
	if r.store == nil {
		return nil
	}
	agents, err := r.store.List(ctx)
	if err != nil {
		return err
	}

	next := make(map[string]*entry, len(agents))

	r.mu.RLock()
	for i := range agents {
		a := agents[i]
		if existing, ok := r.entries[a.ID]; ok && existing.agent.URL == a.URL && existing.agent.Token == a.Token {
			next[a.ID] = &entry{agent: a, client: existing.client}
			continue
		}
		next[a.ID] = &entry{
			agent:  a,
			client: New(a.URL, a.Token, r.clientOpts...),
		}
	}
	r.mu.RUnlock()

	r.mu.Lock()
	r.entries = next
	r.mu.Unlock()

	r.logger.Debug("bonnie: registry reloaded", "count", len(next))
	return nil
}

// Upsert registers (or replaces) a single agent without touching the
// rest of the registry. Used when a service persists a newly registered
// agent and wants it available immediately, without waiting for the
// next poll.
func (r *Registry) Upsert(a Agent) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.entries[a.ID] = &entry{
		agent:  a,
		client: New(a.URL, a.Token, r.clientOpts...),
	}
}

// Remove unregisters an agent by id. Unknown ids are silently ignored.
func (r *Registry) Remove(id string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.entries, id)
}

// Get returns the Client registered for id. The second return value is
// false when no agent with that id is known.
func (r *Registry) Get(id string) (Client, bool) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	e, ok := r.entries[id]
	if !ok {
		return nil, false
	}
	return e.client, true
}

// All returns a snapshot of every registered Client keyed by agent id.
// The returned map is a copy — mutating it has no effect on the
// registry.
func (r *Registry) All() map[string]Client {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make(map[string]Client, len(r.entries))
	for id, e := range r.entries {
		out[id] = e.client
	}
	return out
}

// Agents returns a snapshot of every registered Agent record.
func (r *Registry) Agents() []Agent {
	r.mu.RLock()
	defer r.mu.RUnlock()
	out := make([]Agent, 0, len(r.entries))
	for _, e := range r.entries {
		out = append(out, e.agent)
	}
	return out
}

// Poll runs Health against every registered agent and persists the
// result via RegistryStore.UpdateStatus. Called on each tick of the
// background loop; exposed for tests and for callers that want to
// trigger a refresh on demand.
func (r *Registry) Poll(ctx context.Context) {
	r.mu.RLock()
	snap := make([]*entry, 0, len(r.entries))
	for _, e := range r.entries {
		snap = append(snap, e)
	}
	r.mu.RUnlock()

	for _, e := range snap {
		status := StatusOnline
		if err := e.client.Health(ctx); err != nil {
			status = StatusOffline
			r.logger.Debug("bonnie: registry health check failed",
				"agent", e.agent.Name, "id", e.agent.ID, "error", err)
		}

		now := time.Now().UTC()
		if r.store != nil {
			if err := r.store.UpdateStatus(ctx, e.agent.ID, status, now); err != nil {
				r.logger.Error("bonnie: registry update status failed",
					"agent", e.agent.Name, "id", e.agent.ID, "error", err)
			}
		}

		r.mu.Lock()
		if cur, ok := r.entries[e.agent.ID]; ok {
			cur.agent.Status = status
			cur.agent.LastSeenAt = now
		}
		r.mu.Unlock()
	}
}

// HasOnlineAgent returns nil when at least one registered agent is
// marked online. Useful as a /health sub-check for services that
// require at least one reachable BONNIE host.
func (r *Registry) HasOnlineAgent() error {
	r.mu.RLock()
	defer r.mu.RUnlock()

	if len(r.entries) == 0 {
		return errNoAgentsRegistered
	}
	for _, e := range r.entries {
		if e.agent.Status == StatusOnline {
			return nil
		}
	}
	return errNoOnlineAgents
}

File: src/flag_commons/bonnie/registry.py
"""Agent registry: one client per BONNIE agent plus background health polls."""

from __future__ import annotations

import asyncio
import contextlib
import logging
from collections.abc import Callable
from dataclasses import dataclass, replace
from datetime import datetime, timezone
from typing import Protocol

from flag_commons.bonnie.client import BonnieClient
from flag_commons.bonnie.errors import BonnieUnauthorized

STATUS_ONLINE = "online"
STATUS_OFFLINE = "offline"
STATUS_UNAUTHORIZED = "unauthorized"

DEFAULT_POLL_INTERVAL = 30.0
DEFAULT_RELOAD_INTERVAL = 60.0
DEFAULT_CONCURRENCY = 8

_log = logging.getLogger(__name__)


@dataclass(frozen=True)
class Agent:
    """The shared agent record that flows through :class:`RegistryStore`."""

    id: str
    name: str
    url: str
    token: str = ""
    status: str = STATUS_OFFLINE
    last_seen_at: datetime | None = None
    last_checked_at: datetime | None = None


class RegistryStore(Protocol):
    """Persistence for agents; each service implements it over its own schema."""

    async def list(self) -> list[Agent]: ...

    async def update_status(
        self,
        agent_id: str,
        status: str,
        last_seen_at: datetime | None,
        last_checked_at: datetime,
    ) -> None: ...


class RegistryError(Exception):
    """Raised by :meth:`AgentRegistry.has_online_agent`."""


class NoAgentsRegistered(RegistryError):
    def __init__(self) -> None:
        super().__init__("bonnie: no agents registered")


class NoOnlineAgents(RegistryError):
    def __init__(self) -> None:
        super().__init__("bonnie: no online agents")


ClientFactory = Callable[[str, str], BonnieClient]


@dataclass
class _Entry:
    agent: Agent
    client: BonnieClient


class AgentRegistry:
    """Keeps a :class:`BonnieClient` per agent and polls their ``/health``.

    Differences from the Go registry, all deliberate:

    * the first poll runs immediately at :meth:`start` (Go waited a full
      interval);
    * agents are periodically reloaded from the store (Go never reloaded);
    * polls run concurrently, bounded by ``concurrency`` (Go polled serially);
    * ``last_seen_at`` moves only on success, ``last_checked_at`` on every
      probe;
    * a 401/403 sets ``unauthorized`` instead of ``offline``;
    * calling :meth:`start` twice is harmless.
    """

    def __init__(
        self,
        store: RegistryStore | None,
        *,
        poll_interval: float = DEFAULT_POLL_INTERVAL,
        reload_interval: float = DEFAULT_RELOAD_INTERVAL,
        concurrency: int = DEFAULT_CONCURRENCY,
        client_factory: ClientFactory | None = None,
        logger: logging.Logger | None = None,
    ) -> None:
        self._store = store
        self.poll_interval = poll_interval
        self.reload_interval = reload_interval
        self._semaphore_size = max(1, concurrency)
        self._factory: ClientFactory = client_factory or (
            lambda url, token: BonnieClient(url, token)
        )
        self._log = logger or _log
        self._entries: dict[str, _Entry] = {}
        self._lock = asyncio.Lock()
        self._task: asyncio.Task[None] | None = None

    # --- lifecycle --------------------------------------------------------

    async def start(self) -> None:
        """Reload, poll once, then keep polling and reloading in the background."""
        if self._task is not None and not self._task.done():
            return
        try:
            await self.reload()
        except Exception as exc:  # noqa: BLE001
            self._log.warning("bonnie: registry initial reload failed: %s", exc)
        await self.poll()
        if self.poll_interval > 0:
            self._task = asyncio.create_task(self._loop(), name="bonnie-registry")

    async def stop(self) -> None:
        """Cancel the background loop and close every client."""
        if self._task is not None:
            self._task.cancel()
            with contextlib.suppress(asyncio.CancelledError, Exception):
                await self._task
            self._task = None
        async with self._lock:
            entries = list(self._entries.values())
            self._entries = {}
        for entry in entries:
            await entry.client.aclose()

    async def _loop(self) -> None:
        since_reload = 0.0
        while True:
            await asyncio.sleep(self.poll_interval)
            since_reload += self.poll_interval
            if self.reload_interval > 0 and since_reload >= self.reload_interval:
                since_reload = 0.0
                try:
                    await self.reload()
                except Exception as exc:  # noqa: BLE001
                    self._log.warning("bonnie: registry reload failed: %s", exc)
            try:
                await self.poll()
            except Exception as exc:  # noqa: BLE001 - the loop must survive
                self._log.error("bonnie: registry poll failed: %s", exc)

    # --- membership -------------------------------------------------------

    async def reload(self) -> None:
        """Replace the in-memory set with the store's rows, reusing clients."""
        if self._store is None:
            return
        agents = await self._store.list()
        async with self._lock:
            stale: list[BonnieClient] = []
            next_entries: dict[str, _Entry] = {}
            for agent in agents:
                existing = self._entries.get(agent.id)
                if (
                    existing
                    and existing.agent.url == agent.url
                    and existing.agent.token == agent.token
                ):
                    next_entries[agent.id] = _Entry(agent, existing.client)
                else:
                    if existing:
                        stale.append(existing.client)
                    next_entries[agent.id] = _Entry(
                        agent, self._factory(agent.url, agent.token)
                    )
            for agent_id, entry in self._entries.items():
                if agent_id not in next_entries:
                    stale.append(entry.client)
            self._entries = next_entries
        for client in stale:
            await client.aclose()
        self._log.debug("bonnie: registry reloaded: count=%d", len(next_entries))

    async def upsert(self, agent: Agent) -> None:
        """Register or replace one agent immediately."""
        async with self._lock:
            existing = self._entries.get(agent.id)
            if (
                existing
                and existing.agent.url == agent.url
                and existing.agent.token == agent.token
            ):
                self._entries[agent.id] = _Entry(agent, existing.client)
                return
            self._entries[agent.id] = _Entry(
                agent, self._factory(agent.url, agent.token)
            )
        if existing:
            await existing.client.aclose()

    async def remove(self, agent_id: str) -> None:
        """Unregister an agent. Unknown ids are ignored."""
        async with self._lock:
            entry = self._entries.pop(agent_id, None)
        if entry:
            await entry.client.aclose()

    def get(self, agent_id: str) -> BonnieClient | None:
        entry = self._entries.get(agent_id)
        return entry.client if entry else None

    def all(self) -> dict[str, BonnieClient]:
        return {agent_id: e.client for agent_id, e in self._entries.items()}

    def agents(self) -> list[Agent]:
        return [e.agent for e in self._entries.values()]

    def has_online_agent(self) -> None:
        """Return normally when at least one agent is online, else raise."""
        if not self._entries:
            raise NoAgentsRegistered()
        if not any(e.agent.status == STATUS_ONLINE for e in self._entries.values()):
            raise NoOnlineAgents()

    # --- polling ----------------------------------------------------------

    async def poll(self) -> None:
        """Probe every agent's ``/health`` concurrently and persist the outcome."""
        entries = list(self._entries.values())
        if not entries:
            return
        semaphore = asyncio.Semaphore(self._semaphore_size)

        async def probe(entry: _Entry) -> None:
            async with semaphore:
                await self._probe(entry)

        await asyncio.gather(*(probe(e) for e in entries))

    async def _probe(self, entry: _Entry) -> None:
        now = datetime.now(timezone.utc)
        agent = entry.agent
        try:
            await entry.client.health()
            status = STATUS_ONLINE
        except BonnieUnauthorized as exc:
            status = STATUS_UNAUTHORIZED
            self._log.warning(
                "bonnie: agent rejected token: agent=%s error=%s", agent.name, exc
            )
        except Exception as exc:  # noqa: BLE001 - any failure is "offline"
            status = STATUS_OFFLINE
            self._log.debug(
                "bonnie: agent health check failed: agent=%s error=%s", agent.name, exc
            )

        last_seen = now if status == STATUS_ONLINE else agent.last_seen_at
        if self._store is not None:
            try:
                await self._store.update_status(agent.id, status, last_seen, now)
            except Exception as exc:  # noqa: BLE001
                self._log.error(
                    "bonnie: registry update status failed: agent=%s error=%s",
                    agent.name,
                    exc,
                )
        async with self._lock:
            current = self._entries.get(agent.id)
            if current is not None and current.client is entry.client:
                current.agent = replace(
                    current.agent,
                    status=status,
                    last_seen_at=last_seen,
                    last_checked_at=now,
                )


class BonnieAgentsChecker:
    """Health checker: passes when at least one agent is online.

    Register it as non-critical (``registry.register(checker, critical=False)``)
    so a control plane with no agents yet still reports ready.
    """

    name = "bonnie-agents"

    def __init__(self, registry: AgentRegistry) -> None:
        self._registry = registry

    def check(self) -> None:
        self._registry.has_online_agent()
